### PR TITLE
Alan's work on legend

### PR DESCRIPTION
**Pull request by waharnum**
Creating pull request to incorporate this work into Cindy's FLOE-404 branch.

### CODE REVIEW

**Comment by cindyli** [comment]
Alan, thanks for the good work.

All commit messages need to start with the JIRA number, such as, 

FLOE-404: ...

File: src/js/legend.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        gradeNames: ["gpii.d3ViewComponent", "autoInit"],
+        model: {
+            // dataSet accepts:
+            // 1. an array of objects. Must contain "id", "value" and "label" variables.
+            // Example: [{id: string, value: number, label: string} ... ]
+            dataSet: []
+        },
+        legendOptions: {
+            width: 300,
+            height: 300,
```
Comment (cindyli) [inline src/js/legend.js]: "width" and "height" are un-used. They are probably not required for creating legend.

File: src/js/legend.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            colors: null, // An array of colors for the legend generated for corresponding values of model.dataSet
+            sort: true // Whether or not to sort the data by values when creating the legend
+        },
+        styles: {
+            legend: "gpii-ca-pieChart-legend",
+            table: "gpii-ca-pieChart-legend-table",
+            row: "gpii-ca-pieChart-legend-table-row",
+            legendColorCell: "gpii-ca-pieChart-legend-table-row-legend-color-cell",
+            labelCell: "gpii-ca-pieChart-legend-table-row-legend-label-cell",
+            valueCell: "gpii-ca-pieChart-legend-table-row-legend-value-cell"
```
Comment (cindyli) [inline src/js/legend.js]: Can we shorten names for cell elements? Same as these element defined in selectors block.

File: src/js/legend.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            c.push(d);
+        }
+        return c;
+    };
+
+    gpii.chartAuthoring.pieChart.legend.draw = function(that) {
+
+        var table = that.table,
+            l = that.options.legendOptions,
+            dataSet = gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors(that.model.dataSet,that.color),
```
Comment (cindyli) [inline src/js/legend.js]: This consolidation should use model relay. An example: https://github.com/fluid-project/infusion/blob/master/tests/framework-tests/core/js/DataBindingTests.js#L947-L955

File: src/js/legend.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        });
+
+        addedRows.append("td")
+        .attr({
+            "class": legendValueCellClass
+        })
+        .html(function(d,i) {
+            return d.value;
+        });
+
```
Comment (cindyli) [inline src/js/legend.js:162]: An extra empty line. Same at a few other places.

File: src/js/legend.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        })
+        .html(function(d, i) {
+            return d.label;
+        });
+
+        addedRows.append("td")
+        .attr({
+            "class": legendValueCellClass
+        })
+        .html(function(d,i) {
```
Comment (cindyli) [inline src/js/legend.js]: The infusion coding convention needs a space in btw the keyword "function" and the open bracket, such as "function (".

The same applies to other spots.

File: src/js/legend.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        that.table.append("thead");
+        that.table.append("tbody");
+
+        var thead = that.table.selectAll("thead");
+
+        thead.append("th")
+            .attr({
+                "scope":"col"
+            })
+            .html("Legend");
```
Comment (cindyli) [inline src/js/legend.js]: The static messages to be shown on UI need to go into "strings" option block so they are user configurable.

The doc: http://docs.fluidproject.org/infusion/development/ComponentConfigurationOptions.html#strings
The example: https://github.com/fluid-project/infusion/blob/master/src/components/overviewPanel/js/OverviewPanel.js#L141

Same comment applies to line 184 and 190.

File: tests/html/legend-Tests.html [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        <script src="../lib/infusion/test-core/jqUnit/js/jqUnit.js"></script>
+        <script src="../lib/infusion/test-core/jqUnit/js/jqUnit-browser.js"></script>
+        <script src="../lib/infusion/test-core/utils/js/IoCTestUtils.js"></script>
+
+        <script src="../../src/js/utils.js"></script>
+        <script src="../../src/js/d3ViewComponent.js"></script>
+        <script src="../../src/js/legend.js"></script>
+        <script src="../js/legendTests.js"></script>
+        <style>
+        .gpii-ca-pieChart-table td.gpii-ca-pieChart-table td {
```
Comment (cindyli) [inline tests/html/legend-Tests.html]: Is the test dependent on this css? If not, it should be removed. In the case to add the component css to test files, the css should be in a css file in "src/css/" directory.

File: tests/js/legendTests.js [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    "use strict";
+
+    fluid.registerNamespace("gpii.tests.chartAuthoring");
+
+    fluid.defaults("gpii.tests.chartAuthoring.pieChart.legend", {
+        gradeNames: ["gpii.chartAuthoring.pieChart.legend", "autoInit"],
+        legendOptions: {
+            width: 200,
+            height: 200,
```
Comment (cindyli) [inline tests/js/legendTests.js]: "width" and "height" are un-needed.

File: tests/js/legendTests.js [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+        var d3LabelCells = that.jQueryToD3($(that.locate("valueCell")));
+        d3LabelCells.each(function (d, i) {
+            jqUnit.assertEquals("The data values are applied correctly in the legend", d.value, ($(this).html()));
+        });
+
+    };
+
+    jqUnit.test("Test the legend component created based off an array of objects, unsorted", function () {
+        jqUnit.expect(44);
```
Comment (cindyli) [inline tests/js/legendTests.js]: The functional flow for testing sorted and unsorted legends are same. Can we consolidate into a function to reuse.

**Comment by cindyli** [comment]
Some linting errors.

**Comment by waharnum** [comment]
I'll look at the other issues later this afternoon, but I've just done a rebase and pushed to put the JIRA number at the start of all the commit messages, since that was easy to address.

**Comment by waharnum** [comment]
Looking at the rest of these issues now.

File: src/js/legend.js (hunk 1 above)
Comment (waharnum) [inline src/js/legend.js]: Addressed by d49411c

File: src/js/legend.js (hunk 2 above)
Comment (waharnum) [inline src/js/legend.js]: Addressed by f8ae24d

File: src/js/legend.js (hunk 4 above)
Comment (waharnum) [inline src/js/legend.js:162]: 1344ea7

File: src/js/legend.js (hunk 5 above)
Comment (waharnum) [inline src/js/legend.js]: d6d9a8a

File: src/js/legend.js (hunk 6 above)
Comment (waharnum) [inline src/js/legend.js]: Addressed by f7b712c

File: tests/html/legend-Tests.html (hunk 7 above)
Comment (waharnum) [inline tests/html/legend-Tests.html]: Removed by 19cd02f

File: tests/js/legendTests.js (hunk 9 above)
Comment (waharnum) [inline tests/js/legendTests.js]: They originally had a slightly separate flow, but it became the same on refactoring. Consolidated via 9e700af

File: src/js/legend.js (hunk 3 above)
Comment (waharnum) [inline src/js/legend.js]: Resolved with 6b021fa

File: src/js/legend.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+*/
+
+(function ($, fluid) {
+
+    "use strict";
+
+    fluid.defaults("gpii.chartAuthoring.pieChart.legend", {
+        gradeNames: ["gpii.d3ViewComponent", "autoInit"],
+        strings: {
+          legendColHeading:"Legend",
```
Comment (cindyli) [inline src/js/legend.js]: Indentations are off here as well as some other spots.

File: src/js/legend.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+            dataSet = that.model.dataSet,
+            l = that.options.legendOptions,
+            colors = l.colors,
+            tableClass = that.classes.table;
+
+        if (dataSet.length === 0) {
+            return;
+        }
+
+        that.color = gpii.chartAuthoring.pieChart.legend.getColorScale(colors);
```
Comment (cindyli) [inline src/js/legend.js]: This line can be made declarative into the component defaults block.
An example - https://github.com/fluid-project/infusion/blob/master/src/components/textfieldSlider/js/TextfieldSlider.js#L120-L125

File: src/js/legend.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        })
+        .html(function (d) {
+            return d.value;
+        });
+
+        var removedRows = rows.exit();
+
+        removedRows.remove();
+
+        if(sort === true) {
```
Comment (cindyli) [inline src/js/legend.js]: Can simplify to 

``` javascript
if (sort)
```

Can you add a space btw "if" and the open bracket.

File: src/js/legend.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    };
+
+    // Takes the dataSet array and the color array, and returns a consolidated object array to ease sorting and other operations while keeping colors "correct"
+
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+      var dataSet = model.dataSet;
+      var color = gpii.chartAuthoring.pieChart.legend.getColorScale(model.colors);
+
+        var c = [];
+        for(var i=0; i<dataSet.length; i++) {
```
Comment (cindyli) [inline src/js/legend.js]: fluid.each() is encouraged to use for iterations.

File: src/js/legend.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+    gpii.chartAuthoring.pieChart.legend.getColorScale = function (colors) {
+      return colors ? d3.scale.ordinal().range(colors) : d3.scale.category10();
+    };
+
+    // Takes the dataSet array and the color array, and returns a consolidated object array to ease sorting and other operations while keeping colors "correct"
+
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+      var dataSet = model.dataSet;
+      var color = gpii.chartAuthoring.pieChart.legend.getColorScale(model.colors);
```
Comment (cindyli) [inline src/js/legend.js]: Once this comment https://github.com/cindyli/chartAuthoring/pull/1/files#r38354909 is addressed, "colors" can be directly retrieved form that.colors.

File: src/js/legend.js [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        addedRows
+        .attr({
+            "class": rowClass
+        });
+
+        addedRows
+        .append("td")
+        .attr({
+            "style": function (d) {
+                return "background-color: "+d.color+";";
```
Comment (cindyli) [inline src/js/legend.js]: Can you add spaces before and after "+" sign.

File: tests/js/legendTests.js [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+  }, {
+      id: "id2",
+      value: 20,
+      label: "Three"
+  }
+];
+
+    gpii.tests.chartAuthoring.mouseOverListener = function (data, i, that) {
+        that.mouseOverListenerCalled = true;
+    };
```
Comment (cindyli) [inline tests/js/legendTests.js:176]: Extra spaces in front line 90 - 92.

File: tests/js/legendTests.js [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+
+      // Item added to dataset
+
+      that.applier.change("dataSet", gpii.tests.chartAuthoring.objectArrayAdd);
+      gpii.tests.chartAuthoring.validateLegend(that);
+
+      // Item removed from dataset
+
+      that.applier.change("dataSet", gpii.tests.chartAuthoring.objectArrayRemove);
+      gpii.tests.chartAuthoring.validateLegend(that, false);
```
Comment (cindyli) [inline tests/js/legendTests.js]: gpii.tests.chartAuthoring.validateLegend.validateLegend() doesn't accept the second argument.

**Comment by waharnum** [comment]
Comments should be addressed now.

**Comment by cindyli** [comment]
Thanks for the fix, @waharnum. They look great. 

BTW, there are some linting errors.

File: src/js/legend.js [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+    "use strict";
+
+    fluid.defaults("gpii.chartAuthoring.pieChart.legend", {
+        gradeNames: ["gpii.d3ViewComponent", "autoInit"],
+        strings: {
+            legendColHeading:"Legend",
+            labelColHeading:"Label",
+            valueColHeading:"Value"
+        },
+        model: {
```
Comment (cindyli) [inline src/js/legend.js:22]: Would be nice to add a comment for "dataSetWithColors" path in the model block. It makes easier for other developers to realize as well as for the documentation purpose.

**Comment by cindyli** [comment]
@waharnum, it seems you merged in the branch for the data entry form and committed. Can you clean it up as this branch is only for D3 works. Thanks.

File: src/js/d3ViewComponent.js [hunk 19]
```diff
@@ -56,26 +56,46 @@ https://github.com/gpii/universal/LICENSE.txt
...
+            } else {fluid.set(consolidatedClasses, key, gpii.d3ViewComponent.extractSelectorName(selector));}
+        });
+
+        // 2. For each key in the consolidatedClasses object, split its value into an array of string values separated by spaces,
+        // filter for unique strings, and make a new value based on the array of uniques
+
+        fluid.each(consolidatedClasses, function(classes, key) {
+            var splitClasses = classes.split(" ");
+            // Basic alphabetic sort to improve readability of applied classes and make testing easier
+            splitClasses.sort(function (a,b) {
```
Comment (cindyli) [inline src/js/d3ViewComponent.js]: Infusion's convention for applying selectors to DOM elements are, control selector names, the ones starting with "gpiic-" in the chart authoring context, comes before styling selector names that start with "gpii-". The convention was decided 'cuz usually a DOM element only has one control selector but several styling selectors. Would be nice it the consolidated classes can sustain this convention.

File: src/js/legend.js (hunk 18 above)
Comment (cindyli) [inline src/js/legend.js:22]: Sorry I didn't make it clear. What I meant of adding a comment for "dataSetWithColors" was,

``` javascript
model: {
     dataSet: [],
    // "dataSetWithColors" consolidates the dataset in original order with the array of colors, into a single array of objects
    // dataSetWithColors: []
}
```

When writing component API docs, we normally look at this defaults block. This comment helps to realize there's another model path that's created internally.

File: src/js/pie.js [hunk 20]
```diff
@@ -126,6 +123,12 @@ https://github.com/gpii/universal/LICENSE.txt
                 }
             });
 
+        // Update text values
+        texts.text(function (d) {
```
Comment (cindyli) [inline src/js/pie.js:127]: Thanks for fixing this. A beautiful finding.

File: tests/js/legendTests.js [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+        var hex = parseInt(hexStr.substring(1), 16);
+        var r = (hex & 0xff0000) >> 16;
+        var g = (hex & 0x00ff00) >> 8;
+        var b = hex & 0x0000ff;
+        return "rgb(" + r + ", " + g + ", " + b + ")";
+    };
+
+    gpii.tests.chartAuthoring.testLegendSyncWithModelDataSet = function(that) {
+        var dataSet = that.model.dataSet;
+        //
```
Comment (cindyli) [inline tests/js/legendTests.js]: Stray comment signs.

File: tests/js/legendTests.js [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+        if(that.options.legendOptions.sort) {
+            dataSet.sort(function(a,b) {
```
Comment (cindyli) [inline tests/js/legendTests.js]: Exactly same sorting function used in legend.js. If this piece of code is wrong, the test is not catching it. You probably wanna directly pass in the sorted array value for comparison.

File: tests/js/pieTests.js [hunk 23]
```diff
@@ -133,8 +146,19 @@ https://github.com/gpii/universal/LICENSE.txt
         value: 45
     }];
 
+    gpii.tests.chartAuthoring.objectArrayChangeInPlace = [{
```
Comment (cindyli) [inline tests/js/pieTests.js:149]: Nice catch. Thanks for adding this test scenario.

File: tests/js/legendTests.js (hunk 22 above)
Comment (waharnum) [inline tests/js/legendTests.js]: It's actually that we need to sort the model's dataSet in test in ascending order to compare it against the rendered components, if the sort option is true. At the moment, legend.js leaves the original dataset's ordering alone, but sorts the D3-based elements when sort: true.

It is possible this is overly complicated & we should simply sort the dataSet whenever it's updated if the sort option is true in legend.js. I think my logic was that you may want to preserve original dataset order for other reasons while sorting the visual display.

File: src/js/legend.js (hunk 18 above)
Comment (waharnum) [inline src/js/legend.js:22]: Gotcha, correcting now.

**Comment by cindyli** [comment]
There are some linting errors.

**Pull request merged**